tracing/src/main/java/com/palantir/tracing/CrossThreadSpan.java

```diff
@@ -0,0 +1,146 @@
+import com.palantir.tracing.api.OpenSpan;
+import com.palantir.tracing.api.Span;
+import com.palantir.tracing.api.SpanType;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.concurrent.ThreadSafe;
+
+/** For low-level library code only. */
+@ThreadSafe
+public final class CrossThreadSpan {
+
+    private final String traceId;
+    private final boolean observable;
+    private final AtomicReference<OpenSpan> openSpan;
+
+    private CrossThreadSpan(String traceId, boolean observable, AtomicReference<OpenSpan> openSpan) {
+        this.traceId = traceId;
+        this.observable = observable;
+        this.openSpan = openSpan;
+    }
+
+    @CheckReturnValue
+    public static CrossThreadSpan startSpan(String operation) {
+        Trace trace = Tracer.hasTraceId()
+                ? Tracer.getOrCreateCurrentTrace()
+                : Tracer.createTrace(Observability.UNDECIDED, Tracers.randomId());
+
+        Optional<String> parentSpanId = trace.top().map(OpenSpan::getSpanId);
+        OpenSpan openSpan1 = OpenSpan.of(
+                operation,
+                Tracers.randomId(),
+                SpanType.LOCAL,
+                parentSpanId,
+                trace.getOriginatingSpanId());
+
+        return new CrossThreadSpan(trace.getTraceId(), trace.isObservable(), new AtomicReference<>(openSpan1));
+    }
+
+    /**
+     * Complete the span. This method does not change any thread-local state, so successive operations on this
+     * thread may have different traceIds. If you want to finish the span and continue work related to
+     * this traceId use {@link #sibling}, or if you want to continue work related to this traceId without completing
+     * the span, use {@link #child}.
+     */
+    public void terminate() {
+        OpenSpan inProgress = consumeOpenSpan("CrossThreadSpan may only be terminated once");
+
+        // end the OpenSpan
+        Map<String, String> metadata = Collections.emptyMap();
+        Span span = Tracer.toSpan(inProgress, metadata, traceId);
+
+        Tracer.notifyObservers(span);
+    }
+
+    /**
+     * Finish the cross-thread span and start a new (thread local span) with the given operation name.
+     * NOTE: this is destructive, because any existing threadlocal tracing state is thrown away.
+     */
+    @MustBeClosed
+    public CloseableTracer sibling(String nextOperationName) {
+        OpenSpan inProgress = consumeOpenSpan("Can't call CrossThreadSpan#sibling if span has already been closed");
+
+        Map<String, String> metadata = Collections.emptyMap();
+        Span span = Tracer.toSpan(inProgress, metadata, traceId);
+        Tracer.notifyObservers(span);
+
+        // we throw away the existing thread local state and replace it with our view of the world
+        Tracer.clearCurrentTrace();
+        Tracer.setTrace(Trace.of(observable, traceId));
```

Comment:
I think maybe I'd wrap up the clearCurrentTrace+setTrace into a couple a private method `replaceTraceState` that takes no args, you can reuse it in `child`, too